include/uxr/agent/transport/custom/CustomAgent.hpp

```diff
@@ -0,0 +1,197 @@
+        CustomAgent<EndPoint> *,
+        uint8_t *,
+        InputPacket<EndPoint> &,
+        int,
+        TransportRc &)>;
+    using SendMsgFunction = std::function<bool (
+        CustomAgent<EndPoint> *,
+        uint8_t *,
+        OutputPacket<EndPoint>,
+        TransportRc &)>;
+    using HandleErrorFunction = std::function<bool (
+        CustomAgent<EndPoint> *,
+        TransportRc)>;
+
+    /**
+     * @brief Constructor.
+     * @param middleware_kind   The middleware selected to represent the XRCE entities
+     *                          in the DDS world (FastDDS, FastRTPS, CED...)
+     * @param init_function     Custom user-defined function, called during initialization.
+     * @param fini_function     Custom user-defined function, called upon agent's destruction.
+     * @param recv_msg_function Custom user-defined function, called when receiving some data.
+     * @param send_msg_function Custom user-defined function, called when sending some information.
+     * @param error_function    Custom user-defined function, called to recover from an error.
+     */
+    CustomAgent(
+            Middleware::Kind middleware_kind,
+            InitFunction&& init_function,
+            FiniFunction&& fini_function,
+            RecvMsgFunction&& recv_msg_function,
+            SendMsgFunction&& send_msg_function,
+            HandleErrorFunction&& error_function)
+        : Server<EndPoint>(middleware_kind)
+        , custom_init_func_(std::move(init_function))
+        , custom_fini_func_(std::move(fini_function))
+        , custom_recv_msg_func_(std::move(recv_msg_function))
+        , custom_send_msg_func_(std::move(send_msg_function))
+        , custom_handle_error_func_(std::move(error_function))
+    {
+    }
+
+    /**
+     * @brief Destructor.
+     */
+    ~CustomAgent() final
+    {
+        try
+        {
+            this->stop();
+        }
+        catch(std::exception& e)
+        {
+            UXR_AGENT_LOG_CRITICAL(
+                UXR_DECORATE_RED("Error stopping custom agent server"),
+                "exception: {}",
+                e.what());
+        }
+    }
+
+    /**
+     * @brief Helper method to retrieve the custom agent's internal buffer size.
+     *        Useful for some read/write operations.
+     * @return Buffer size, in bytes.
+     */
+    uint16_t buffer_size() const
+    {
+        return static_cast<uint16_t>(SERVER_BUFFER_SIZE);
+    }
+
+private:
+    /**
+     * @brief Override virtual Server operations.
+     */
+    inline bool init() final
+    {
+        return custom_init_func_(this);
+    }
+
+    inline bool fini() final
+    {
+        return custom_fini_func_(this);
+    }
+
+#ifdef UAGENT_DISCOVERY_PROFILE
+    inline bool init_discovery(
+            uint16_t /*discovery_port*/) final
+    {
+        return false;
+    }
+
+    inline bool fini_discovery() final
+    {
+        return false;
+    }
+#endif // UAGENT_DISCOVERY_PROFILE
+
+#ifdef UAGENT_P2P_PROFILE
```

Comment:
Could you explain this a bit? Is this the old P2P or the new approach? Please clarify

Reply:
This is related to the old approach. If this flag is enabled, `eprosima::uxr::Server` class will declare this virtual methods and thus `CustomAgent` class must override them. I just returned false since I don't think this is something users would need to set in their custom agent implementation, but we can change it anytime.

Reply:
Cool